### PR TITLE
MODSI-8, database design, add match_point table. 

### PR DESCRIPTION
**Pull request by nielserik**
The PR implements a change to the table structure of the shared index - from having one table, BIB_RECORD, with a fixed number of match value columns, to a normalized, two-table structure, which still has a BIB_RECORD table, but now without any match value columns, and then a MATCH_POINT table to hold those match values instead:

MATCH_POINT
   (BIB_RECORD_ID,
    MATCH_TYPE_ID,
    MATCH_VALUE)

This change is necessary primarily because it allows for an unlimited number of match values of a given type per bib record, for example an unbounded number of ISBNs per bib record, and secondly because it allows for an unbounded number of different match types (defined by match_type_id) with no need to add additional columns to BIB_RECORD for each new match type that we might want to add later.

The change in database structure does not affect the APIs for ingestion or retrieval. It does of course require internal changes to the API implementations, in order to populate and query the table structure.  So in this PR, ingestion is implemented by a trigger on BIB_RECORD that invokes a stored function to update MATCH_POINT in case of inserts or updates to BIB_RECORD (deletion of records in BIB_RECORD are simply cascaded to MATCH_POINT and are thus not part of the stored function). The trigger approach is simple in this case, because the MATCH_POINT table is entirely dependent on the BIB_RECORD table. This implementation will work but it is for proof of concept (that the APIs don't need to change), meaning that, if desired, the ingestion logic could be implemented in Vert.X instead. 

As for batch updates and transactions, this implementation means, that BIB_RECORD can be updated in batch or with longer transactions spanning multiple insert, update or delete operations on BIB_RECORD. The subsequent, "inline" updates to MATCH_POINT would be part of those transactions and be committed and rolled back with them. 

The tables are not meant to be complete and several details, like the use of identifiers for match-types etc, are preliminary. They are merely chosen to be able to use 'inventory' JSON schema that we have at hand from the old SI.

### CODE REVIEW

File: src/main/java/org/folio/reshare/index/storage/Storage.java [hunk 1]
```diff
@@ -50,23 +61,62 @@ public Storage(Vertx vertx, String tenant) {
...
+                + " CONSTRAINT match_point_fk_bib_record FOREIGN KEY "
+                + "                (bib_record_id) REFERENCES " + bibRecordTable
+                + " ON DELETE CASCADE,"
+                + " CONSTRAINT match_point_fk_type FOREIGN KEY "
+                + "                (match_type_id) REFERENCES " + matchTypeTable + ")",
+            "CREATE UNIQUE INDEX IF NOT EXISTS match_point_idx ON " + matchPointTable
+                + " (match_type_id, match_value, bib_record_id)",
+            "CREATE INDEX IF NOT EXISTS match_point_bib_id_idX ON " + matchPointTable
+                + " (bib_record_id)",
+            "CREATE OR REPLACE FUNCTION " + matchPointFunction
```
Comment (skoczko) [inline src/main/java/org/folio/reshare/index/storage/Storage.java:96]: Let's avoid hard-coding any specific match key methods here. These should be "pluggable" so let's factor them out to a separate class or classes.

File: src/main/java/org/folio/reshare/index/storage/Storage.java [hunk 2]
```diff
@@ -36,6 +43,10 @@
   public Storage(Vertx vertx, String tenant) {
     this.pool = TenantPgPool.pool(vertx,tenant);
     this.bibRecordTable = pool.getSchema() + ".bib_record";
+    this.matchTypeTable = pool.getSchema() + ".match_type";
```
Comment (skoczko) [inline src/main/java/org/folio/reshare/index/storage/Storage.java:46]: We should align table names with the OpenAPI spec:

* `match_type` -> `MATCH_KEY_CONFIGS`
* `match_point` -> `MATCH_KEY_VALUES`

would be one way. But we can also go the other way and align the OpenAPI spec. @adamdickmeiss @nielserik preferences?

File: src/main/java/org/folio/reshare/index/storage/Storage.java [hunk 3]
```diff
@@ -24,8 +24,15 @@
 public class Storage {
   private static final Logger log = LogManager.getLogger(Storage.class);
 
+  private static final String IDENTIFIER_TYPE_ID_ISBN = "8261054f-be78-422d-bd51-4ed9f33c3422";
```
Comment (skoczko) [inline src/main/java/org/folio/reshare/index/storage/Storage.java:27]: These constants should be factored out to individual match key method implementation classes.

**Review by skoczko — CHANGES_REQUESTED** [review]

File: src/main/java/org/folio/reshare/index/storage/Storage.java [hunk 4]
```diff
@@ -50,23 +61,62 @@ public Storage(Vertx vertx, String tenant) {
...
+                + "   IF (TG_OP = 'UPDATE' OR TG_OP = 'INSERT') THEN"
+                + "     INSERT INTO " + matchPointTable
+                + "       (bib_record_id, match_type_id, match_value)"
+                + "     SELECT NEW.id, '" + MATCH_TYPE_ID_ISBN + "', identifier->>'value'"
+                + "     FROM jsonb_array_elements("
+                + "           ((NEW.inventory->>'instance')::JSONB->>'identifiers')::JSONB) "
+                + "            AS identifier"
+                + "            WHERE identifier->>'identifierTypeId'='"
+                +                    IDENTIFIER_TYPE_ID_ISBN + "';"
+                + "   END IF;"
```
Comment (adamdickmeiss) [inline src/main/java/org/folio/reshare/index/storage/Storage.java:112]: I'm not much fund of triggers or SQL functions.. 1: I don't understand it (:-) .. 2: They are difficult to unit test.. 3. I think parsing of JSON should be done in Java code and knowledge about UUIDs in inventory should definitely not live in an SQL function.. 4: upgrading problems with triggers/functions.. 

So if the logic is "universal", a trigger MAY be good.. But I don't think in this case, that applies..

**Comment by nielserik** [comment]
The only thing that matters in this PR is the table structure. 

How tables and columns are named, and with what coding languages the population of the tables is implemented is secondary. 

So if it provides clarity to have other names for database elements or the API schema elements, that's totally fine. 

The PR does not suggest that ingestion should be implemented with stored functions, on the contrary. As pointed out in the description: "This implementation will work but it is for proof of concept (...) if desired, the ingestion logic could be implemented in Vert.X instead."  Writing 20 lines of plpgsql code just happened to be the quickest way to implement a demo of it. It could be any other language or framework of your choice of course.   

Also, please note that the Inventory UUID is only present in this PoC to make it work with the test data we have at hand (from the ReShare harvest jobs). Those data happen to be FOLIO Inventory data so the code must pick ISBN numbers by the identifier UUID of ISBNs in Inventory, that's all. In the end -- I would expect -- no Inventory UUIDs or codes should live in the code of mod-shared-index. 

Once a schema is established for that incoming JSON (and I would not expect that to be an Inventory schema), the ingestion code will need to implement its own way of finding the ISBNs in the object. Perhaps the ISBNs will arrive as a JSON array of ISBN values in that JSON, which would greatly simply find that parse. 

What matters is, that the values end up in the table MATCH_POINT,  aka MATCH_KEY_VALUES, using logic equivalent to what is demonstrated in the "match_points()" function.

**Comment by nielserik** [comment]
The table structure is important because it essential for fast and flexible retrieval of clusters of holdings. Perhaps it would be helpful to go through some examples of retrieval logic.

**Review by adamdickmeiss — APPROVED** [review]
I don't mind this merged now. The match key must be computed client side obviously later.

**Pull request closed without merging**